### PR TITLE
Openshift Registry projects/users/groups listing

### PR DESCRIPTION
**Pull request by sub-mod**
Basic listing of projects/users/groups for cluster-admin

### CODE REVIEW

File: pkg/registry/scripts/kube-client-cockpit.js [hunk 1]
```diff
@@ -393,6 +393,11 @@
 
                         } else if (!blocked) {
                             console.warn("watching kube " + path + " didn't block");
+                            if (path === "/oapi/v1/projects") {
+                                response.status = 401;
+                                response.message = msg;
+                                loadFinish(response);
```
Comment (stefwalter) [inline pkg/registry/scripts/kube-client-cockpit.js]: What's going on here? Why are we setting this to 401 (Unauthorized)

File: pkg/registry/scripts/main.js [hunk 2]
```diff
@@ -93,7 +93,7 @@
             /* Curtains related logic */
             function connect() {
                 $scope.curtains = { };
-                loader.watch("namespaces").then(function() {
+                discover().then(function() {
```
Comment (sub-mod) [inline pkg/registry/scripts/main.js]: This code was added because a project-admin cannot list namespaces, 
so to let project-admin access UI we check for discover() to pass

Comment (stefwalter) [inline pkg/registry/scripts/main.js]: Makes sense.

File: pkg/registry/scripts/projects.js [hunk 3]
```diff
@@ -36,8 +38,43 @@
...
+
+            $scope.groups = function() {
+                return select().kind("Group");
+            };
+
+            $scope.projects = function() {
+                return select().kind("Project");
+            };
+
+            $scope.formatMembers = function format_members(members, kind) {
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Could you add a unit test for this function?

File: pkg/registry/styles/projects.less [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+table.projects-listing {
+
+    #add-button-icon {
+        float: right;
+        color: #0099d3;
+        padding: 10px 0px 10px;
+    }
+
+    #add-project-icon {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: I don't see this being used anywhere.

File: pkg/registry/styles/projects.less [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        padding: 30px 0px 10px;
+    }
+
+    #add-user-icon {
+        float: right;
+        color: #0099d3;
+        padding: 30px 0px 10px;
+    }
+
+    tbody:hover {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: Why is this necessary? Shouldn't this be done by the listing stuff in cockpit.css?

File: pkg/registry/styles/projects.less [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        background-color: #d4edfa;
+    }
+
+    thead {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: Why is this necessary, shouldn't this be done by the listing stuff in cockpit.css?

File: pkg/registry/views/projects-page.html [hunk 7]
```diff
@@ -1,6 +1,84 @@
...
-</filter-bar>
-
 <div>
-    <h1>Projects place holder</h1>
+    <table listing-table class="listing projects-listing">
+        <thead ng-if="!quiet">
+            <tr>
+                <td colspan="4">
+                    <span>
+                    <div id="add-button-icon">
```
Comment (stefwalter) [inline pkg/registry/views/projects-page.html]: A <span> is a inline element. We shouldn't need to wrap <div> and <h3> in a span.

File: pkg/registry/views/projects-page.html [hunk 8]
```diff
@@ -1,6 +1,84 @@
...
+            </tr>
+            <tr class="listing-panel" ng-if="selected(id)">
+                
+            </tr>
+        </tbody>
+
+        <thead ng-if="!quiet">
+            <tr>
+                <td colspan="4">
+                    <span>
```
Comment (stefwalter) [inline pkg/registry/views/projects-page.html]: Span should be removed.

**Comment by stefwalter** [comment]
Nice work. Mostly looks good. Just a few review issues

File: pkg/registry/scripts/projects.js [hunk 9]
```diff
@@ -34,10 +35,54 @@
...
 
+    .factory("projectUtil", [
+        function() {
+            return {
+                formatMembers: function (members, kind) {
+                    var mlist = "";
+                    if (!members || members.length === 0)
+                        return mlist;
+                    if (members.length <= 3) {
+                        for (var i = members.length - 1; i >= 0; i--) {
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: We shouldn't be declaring variables inside of an if statement. Unfortunately that's how javascript is. Since |i| is valid across the whole function, lets keep it declared outside of any if statement or other scope element.

File: pkg/registry/scripts/projects.js [hunk 10]
```diff
@@ -34,10 +35,54 @@
...
     .controller('ProjectsCtrl', [
         '$scope',
-        function($scope) {
-            /* nothing here yet */
+        'kubeLoader',
+        'kubeSelect',
+        '$modal',
+        'projectUtil',
+        function($scope, loader, select, $modal, util) {
+            loader.watch(["users"]);
```
Comment (stefwalter) [inline pkg/registry/scripts/projects.js]: Watch just accepts a string. not an array.

File: pkg/registry/styles/projects.less [hunk 11]
```diff
@@ -0,0 +1,44 @@
+
+@tag-border: @link-color;
+
+table.projects-listing {
+
+	tr {
```
Comment (stefwalter) [inline pkg/registry/styles/projects.less]: Indentation needs to be fixed here.

File: pkg/registry/tests/test-projects.html [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+-->
+<html>
+<head>
+    <meta charset="utf-8">
+    <title>Image filter tests</title>
```
Comment (stefwalter) [inline pkg/registry/tests/test-projects.html]: title needs fixing?

**Comment by stefwalter** [comment]
Wonderful. Merged!
